modin/core/dataframe/pandas/metadata/dtypes.py

```diff
@@ -12,13 +12,518 @@
+                and remaining_dtype is None
+                and len(self._known_dtypes) > 0
+            ):
+                self._schema_is_known = len(cols_with_unknown_dtypes) == 0
+
+        self._know_all_names: bool = know_all_names
+        # a common dtype for columns that are not present in 'known_dtypes' nor in 'cols_with_unknown_dtypes'
+        self._remaining_dtype: Optional[np.dtype] = remaining_dtype
+        self._parent_df: Optional["PandasDataframe"] = parent_df
+        if columns_order is None:
+            self._columns_order: Optional[dict[int, IndexLabel]] = None
+            # try to compute '._columns_order' using 'parent_df'
+            self.columns_order
+        else:
+            if remaining_dtype is not None:
+                raise ValueError(
+                    "Passing 'columns_order' and 'remaining_dtype' is ambiguous. You have to manually "
+                    + "complete 'known_dtypes' using the information from 'columns_order' and 'remaining_dtype'."
+                )
+            elif not self._know_all_names:
+                raise ValueError(
+                    "Passing 'columns_order' and 'know_all_names=False' is ambiguous. You have to manually "
+                    + "complete 'cols_with_unknown_dtypes' using the information from 'columns_order' "
+                    + "and pass 'know_all_names=True'."
+                )
+            elif len(columns_order) != (
+                len(self._cols_with_unknown_dtypes) + len(self._known_dtypes)
+            ):
+                raise ValueError(
+                    "The length of 'columns_order' doesn't match to 'known_dtypes' and 'cols_with_unknown_dtypes'"
+                )
+            self._columns_order: Optional[dict[int, IndexLabel]] = columns_order
+
+    def update_parent(self, new_parent: "PandasDataframe"):
+        """
+        Set new parent dataframe.
+
+        Parameters
+        ----------
+        new_parent : PandasDataframe
+        """
+        self._parent_df = new_parent
+        LazyProxyCategoricalDtype.update_dtypes(self._known_dtypes, new_parent)
+        # try to compute '._columns_order' using 'new_parent'
+        self.columns_order
+
+    @property
+    def columns_order(self) -> Optional[dict[int, IndexLabel]]:
+        """
+        Get order of columns for the described dataframe if available.
+
+        Returns
+        -------
+        dict[int, IndexLabel] or None
+        """
+        if self._columns_order is not None:
+            return self._columns_order
+        if self._parent_df is None or not self._parent_df.has_materialized_columns:
+            return None
+
+        self._columns_order = {i: col for i, col in enumerate(self._parent_df.columns)}
+        # we got information about new columns and thus can potentially
+        # extend our knowledge about missing dtypes
+        if len(self._columns_order) > (
+            len(self._known_dtypes) + len(self._cols_with_unknown_dtypes)
+        ):
```

Comment:
Why is the `"The length of 'columns_order' doesn't match to 'known_dtypes' and 'cols_with_unknown_dtypes'"` exception thrown in the constructor when the lengths do not match, but here there is a recalculation?

Reply:
Because if you know `columns_order` beforehand, you can and IMHO must complete `known_dtypes` and `cols_with_unk...` with the information you have at your own. The `DtypesDescriptor` constructor's parameter matrix is already quite complicated, and I didn't want to bring yet another degree of freedom like "oh, and you also can provide some incomplete argument and we'll infer everything magically from the rest of the arguments". I rather wanted to put as many limitations as possible in order to simplify the constructor's logic and avoid having potentially missed and unprocessed cases because of the bloated parameters matrix.